### PR TITLE
Making fastjet 3.1.0 more thread-safe, Take 2

### PR DESCRIPTION
**Pull request by rappoccio**
Revisiting #6 . Here I started with the CMS patches to 3.0.3 and tried to follow Chris's example more closely.

### CODE REVIEW

**Comment by rappoccio** [comment]
@Dr15Jones and @ktf  : any hope of getting this in? It's really rather critical for 73x.

**Comment by Dr15Jones** [comment]
Hopefully I'll get a chance to look at this on Thursday.

**Comment by Dr15Jones** [comment]
@ktf 
Giulio,
What branch should this pull request actually be against? I thought we normally kept the 'master' branch in synch with the official 3rd party repository.

**Comment by ktf** [comment]
Yes, this should definitely go on top of `cms/v3.1.0`.

File: src/ClusterSequenceActiveArea.cc [hunk 1]
```diff
@@ -38,9 +38,22 @@
...
+#include <fstream>
+#if __cplusplus >= 201103L
+#include<atomic>
+#endif
 
 FASTJET_BEGIN_NAMESPACE      // defined in fastjet/internal/base.hh
 
+//CMS change: use std::atomic for thread safety.
+//   Change not endorsed by fastjet collaboration
+#if __cplusplus >= 201103L
```
Comment (Dr15Jones) [inline src/ClusterSequenceActiveArea.cc]: Was this from me? This still doesn't make calling cout our cerr thread safe. The messages will still get all garbled up.

File: src/LazyTiling25.cc [hunk 2]
```diff
@@ -31,11 +31,26 @@
...
 using namespace std;
 
 
 FASTJET_BEGIN_NAMESPACE      // defined in fastjet/internal/base.hh
 
+
+//CMS change: use std::atomic for thread safety.
+//   Change not endorsed by fastjet collaboration
+#if __cplusplus >= 201103L
+static std::atomic<std::ostream*> _safe_cerr{& std::cerr};
```
Comment (Dr15Jones) [inline src/LazyTiling25.cc]: Again, this doesn't help with thread safety.

**Comment by Dr15Jones** [comment]
Changing a `std::cout` to `std::atomic<std::ostream*> new_cout{ & std::cout }` does not solve the underlying thread safety problem. It is the calling of `cout``(or``cerr``) which is not thread safe. Now in practice, I don't think calling them simultaneously on different threads actually causes a crash, it is just that the messages become garbled.
So I suggest switching back to using``std::cout``and``std::cerr`` since at least it keeps the code closer to the original.

**Comment by Dr15Jones** [comment]
I see only three 'solutions' to the `std::cout``and``std::cerr`` problems.
1) ignore them and just live with messed up output
2) comment them out since they are just filling our logfiles with stuff no one looks at
3) create a way to allow us to 'plug-in' our MessageLogger as an 'ostream' to which they send messages.

**Comment by Dr15Jones** [comment]
Sal,
As a test, please run valgrind's helgrind tool on a cmsRun job which would call this code. I suggest only 10 events or so else you will have to wait an enormous amount of time. The command I use is

```
valgrind --tool=helgrind --num-callers=40 --suppressions=$CMSSW_RELEASE_BASE/src/Utilities/ReleaseScripts/data/cms-valgrind-helgrind.supp cmsRun <your config file>
```

Then look for any call stacks containing code from fastjet.

**Comment by rappoccio** [comment]
Thanks Chris! I think option 2 is best here, the output is going to be ignored anyway. 

I will run the helgrind tool today.

**Comment by Dr15Jones** [comment]
So I did a google check and found that C++11 does guarantee that cout and cerr won't crash when multiple threads write to them but they will garble the order of characters. The actual quote was:

Concurrent access to a synchronized (§27.5.3.4) standard iostream object’s formatted and unformatted input (§27.7.2.1) and output (§27.7.3.1) functions or a standard C stream by multiple threads shall not result in a data race (§1.10). [ Note: Users must still synchronize concurrent use of these objects and streams by multiple threads if they wish to avoid interleaved characters. — end note ]

**Comment by rappoccio** [comment]
I did the checks, and there were no instances of fastjet in the helgrind output.

I think (hope) we're good to go.

**Comment by Dr15Jones** [comment]
The only change needed is we need a pull request against the cms/v3.1.0 branch, not the master branch.

**Comment by rappoccio** [comment]
Whoops! My fault. I will close this and add a new PR.

**Pull request closed without merging**